### PR TITLE
enable `-Wbad-function-cast` compiler warning

### PR DESCRIPTION
**Pull request by fabiangreffrath**

### CODE REVIEW

File: src/i_3dsound.c [hunk 1]
```diff
@@ -24,7 +24,7 @@
 #include "r_main.h"
 #include "p_setup.h"
 
-#define FIXED_TO_ALFLOAT(x) ((ALfloat)((double)(x) / FRACUNIT))
+#define FIXED_TO_ALFLOAT(x) ((ALfloat)((x) / (double)FRACUNIT))
```
Comment (fabiangreffrath) [inline src/i_3dsound.c:27]: This is now mostly the same as the `FIXED2DOUBLE()` macro in `m_fixed.h`.

File: src/p_extnodes.c [hunk 2]
```diff
@@ -195,11 +195,13 @@ mapformat_t P_CheckMapFormat(int lumpnum)
 fixed_t P_GetOffset(vertex_t *v1, vertex_t *v2)
 {
     fixed_t dx, dy;
+    double rf;
     fixed_t r;
 
     dx = (v1->x - v2->x) >> FRACBITS;
     dy = (v1->y - v2->y) >> FRACBITS;
-    r = (fixed_t) (sqrt(dx * dx + dy * dy)) << FRACBITS;
+    rf = sqrt(dx * dx + dy * dy);
```
Comment (fabiangreffrath) [inline src/p_extnodes.c:203]: Could as well get replaced with the all-float version from PrBoom+:

https://github.com/kraflab/dsda-doom/blob/18e1125afb2b572ccdb3e99021ff2c95c06c3edf/prboom2/src/p_setup.c#L519-L527

File: src/p_saveg.c [hunk 3]
```diff
@@ -257,14 +257,17 @@ static void saveg_write_mapthing_t(mapthing_t *str)
...
     // struct thinker_s* prev;
     str->prev = saveg_readp();
 
     // struct thinker_s* next;
     str->next = saveg_readp();
 
     // think_t function;
-    str->function.v = (actionf_v)(intptr_t)saveg_readp();
+    func = saveg_readp();
+    str->function.v = (actionf_v)(intptr_t)func;
```
Comment (rfomin) [inline src/p_saveg.c:270]: All of these addresses don't work anyway. Perhaps we should redo this for clarity:
```c
saveg_readp();
str->function.v = NULL;
```

Comment (fabiangreffrath) [reply src/p_saveg.c:270]: Yes. I wasn't sure and wanted to get rid of the warning in the first place, but this is even better.

File: src/i_winmusic.c [hunk 4]
```diff
@@ -1462,7 +1462,7 @@ static void FillBuffer(void)
 // multimedia functions from inside the callback function, as doing so can
 // cause a deadlock." We use a thread to avoid possible deadlocks.
 
-static DWORD WINAPI PlayerProc(void)
+static DWORD WINAPI PlayerProc(void *p)
```
Comment (rfomin) [inline src/i_winmusic.c:1465]: I suggest using `static DWORD WINAPI PlayerProc(LPVOID lpParam)` for consistency with Win API.

**Review by rfomin — COMMENTED** [review]
I've got this error on MSYS2 CLANG64:
```
C:/develop/woof/src/i_3dsound.c:175:56: error: cast from 'void (*)(void)' to 'actionf_v' (aka 'void
(*)()') converts to incompatible function type [-Werror,-Wcast-function-type-strict]
  175 |     src->point_source = (source->thinker.function.v != (actionf_v)P_DegenMobjThinker &&
      |                                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

**Comment by fabiangreffrath** [comment]
> I've got this error on MSYS2 CLANG64:
> 
> ```
> C:/develop/woof/src/i_3dsound.c:175:56: error: cast from 'void (*)(void)' to 'actionf_v' (aka 'void
> (*)()') converts to incompatible function type [-Werror,-Wcast-function-type-strict]
>   175 |     src->point_source = (source->thinker.function.v != (actionf_v)P_DegenMobjThinker &&
>       |                                                        ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> 1 error generated.
> ```

I'm too far away from a C compiler right now. Does the following help or produce even more problems?

```patch
--- a/src/d_think.h
+++ b/src/d_think.h
@@ -22,7 +22,7 @@
 #ifndef __D_THINK__
 #define __D_THINK__

-typedef void (*actionf_v)();
+typedef void (*actionf_v)(void);
 typedef void (*actionf_p1)(void *);
 typedef void (*actionf_p2)(void *, void *);

```

**Comment by rfomin** [comment]
> Does the following help or produce even more problems?

"fatal error: too many errors emitted, stopping now [-ferror-limit=]" with MSYS2 CLANG64

**Comment by fabiangreffrath** [comment]
> "fatal error: too many errors emitted, stopping now [-ferror-limit=]" with MSYS2 CLANG64

Wow! 🤯

**Comment by rfomin** [comment]
I fixed it with 
```diff
--- a/src/i_3dsound.c
+++ b/src/i_3dsound.c
@@ -172,7 +172,7 @@ static void CalcDistance(const mobj_t *listener, const mobj_t *source,
     CalcHypotenuse(adx, ady, &distxy);

     // Treat monsters and projectiles as point sources.
-    src->point_source = (source->thinker.function.v != (actionf_v)P_DegenMobjThinker &&
+    src->point_source = (source->thinker.function.v != P_DegenMobjThinker &&
                         source->info && source->actualheight);

     if (src->point_source)
```

But got next one:
```
C:/develop/woof/src/i_oalmusic.c:304:28: error: cast from 'void (*)(unsigned int) __attribute__((nor
eturn))' to 'pfnSDL_CurrentEndThread' (aka 'void (*)(unsigned int)') converts to incompatible functi
on type [-Werror,-Wcast-function-type-strict]
  304 |     player_thread_handle = SDL_CreateThread(PlayerThread, NULL, NULL);
      |                            ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
C:/msys64/clang64/include/SDL2/SDL_thread.h:144:119: note: expanded from macro 'SDL_CreateThread'
  144 | #define SDL_CreateThread(fn, name, data) SDL_CreateThread(fn, name, data, (pfnSDL_CurrentBeg
inThread)SDL_beginthread, (pfnSDL_CurrentEndThread)SDL_endthread)
      |
                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```

Eh, maybe it's not worth it?

**Comment by fabiangreffrath** [comment]
> Eh, maybe it's not worth it?

Wow, it even considers different function attributes as incompatible. Looks like this would need to be fixed in SDL, right?

**Comment by fabiangreffrath** [comment]
It's part of `-Wextra`, so not *that* ominious.

**Comment by fabiangreffrath** [comment]
> I don't think this is a bug. Perhaps this warning is disabled by default for a reason.

I wonder why it doesn't trigger in the MSYS2 UCRT64 build?

**Comment by rfomin** [comment]
> I wonder why it doesn't trigger in the MSYS2 UCRT64 build?

It uses GCC.

**Comment by fabiangreffrath** [comment]
> It uses GCC.

Ah, sure. 🤦 

Well, I'm not sure how to proceed from here. My intention was to get rid of some quirks and bad practise, but it seems that I would have to add even more of this to silence the warnings that I asked for. 😞

**Comment by fabiangreffrath** [comment]
Could you check if it compiles with the latest commit? Even if it does, I'm probably going to remove the  `-Wcast-function-type` warning altogether. However, I'd like to keep some, if not all of the changes that this experiment triggered.

**Review by rfomin — APPROVED** [review]
Works in MSYS2 CLANG64.

**Comment by fabiangreffrath** [comment]
> Works in MSYS2 CLANG64.

Thanks!

**Pull request merged**